### DISCUSSION TITLE
[질문] AuthActivity에서 자동로그인 관련 로직이 이해가 안가요! (사진 첨부)

### DISCUSSION THREAD

**Discussion by wonseok97** [TEAM-DAME/Dame-Android#42] [2. Q&A] [answered]
<img width="743" alt="image" src="https://user-images.githubusercontent.com/118143620/201584727-07a6e80e-5bf6-4ab0-92a8-92e463bc08c6.png">

**Comment by onseok** [comment, marked as answer]
- `@AndroidEntryPoint` 어노테이션은 [Hilt](https://developer.android.com/training/dependency-injection/hilt-android?hl=ko)에서 제공하는 어노테이션입니다. 해당 어노테이션이 있는 컴포넌트에 `Hilt`는 종속 항목 삽입 (Dependency Injection)을 제공할 수 있습니다. 종속 항목 삽입이 필요한 Activity나 Fragment 클래스에는  전부 해당 주석을 추가해줘야 한다고 보면 됩니다.

- `by Viemodels()`를 사용하면 기존의 `ViewModelProvider`를 사용하지 않고 필요한 `ViewModel`을 지연 생성할 수 있습니다.

- `Android 12`에서는 `Splash Screen`이 강제되어 따로 커스텀하지 않는다면 앱 아이콘이 디폴트로 스플래시 화면에 나타나게 됩니다. 또 기존의 방법대로, 가령 `styles에 windowBackground 속성을 넣고, 스플래시를 보여줄 화면에 테마로 지정하는 방법` 또는 `액티비티나 프래그먼트에 딜레이를 주어 직접 스플래시 화면을 보여주는 방법`을 통해 `Splash Screen`을 구현한다면, 앱이 `hot start`상태에 있더라도 다시 스플래시 스크린을 보여줌으로써 대기시간이 길어져 사용자들에게 불편함을 초래할 수 있습니다. 더 자세한 레퍼런스를 참고하고 싶으시다면 [Splash Screens](https://developer.android.com/develop/ui/views/launch/splash-screen) 공식 문서를 읽어보시길 바랍니다.

- `autoLogin()` 메서드는 [ViewTreeObserver.OnPreDrawListener](https://developer.android.com/reference/android/view/ViewTreeObserver.OnPreDrawListener)를 사용하여 특정한 조건에 맞을 때까지 (현재 코드에서는 `autoLoginState` 변수가 초기화 될 때까지) 앱이 첫 번째 화면을 그리는 것을 지연시킵니다. 만약 `::autoLoginState.isInitialized` 다음의 조건문이 참이라면, `autoLoginState` 변수가 초기화된 상태라는 뜻이고, 해당 변수가 자동 로그인 성공을 뜻하는 `AUTO_LOGIN_SUCCESS`이라면, `로그인` 화면을 건너뛰고 바로 `메인`화면으로 넘어가고, 자동 로그인 실패를 뜻하는 `AUTO_LOGIN_FAILURE`이라면 (여기서는 else 분기), `로그인`화면으로 넘어갑니다. 참고로 replace()는 커스텀된 메서드인데 이는 액티비티 위에 있는 [FragmentContainerView](https://developer.android.com/reference/androidx/fragment/app/FragmentContainerView)에서 프래그먼트 간 화면 전환을 담당하는 역할을 합니다. 자세한 구현체는 아래에 첨부하겠습니다. 

```
inline fun <reified T : Fragment> AppCompatActivity.replace(@IdRes frameId: Int) {
    supportFragmentManager.commit {
        replace<T>(frameId)
        setReorderingAllowed(true)
    }
}
```

- `checkAutoLoginState()`메서드는 `EncryptedSharedPreference`를 통해 기기 내에 저장된 액세스 토큰(JWT 토큰)을 읽어와서 `autoLoginState` 변수 값을 셋팅하는 메서드입니다.
